File: src/jvm/clojure/lang/platform/PrimitiveArray.java
package clojure.lang.platform;

public final class PrimitiveArray {

  public static boolean[] castToBooleans(Object arr) {
    return (boolean[]) arr;
  }

  public static byte[] castToBytes(Object arr) {
    return (byte[]) arr;
  }

  public static char[] castToChars(Object arr) {
    return (char[]) arr;
  }

  public static short[] castToShorts(Object arr) {
    return (short[]) arr;
  }

  public static float[] castToFloats(Object arr) {
    return (float[]) arr;
  }

  public static double[] castToDoubles(Object arr) {
    return (double[]) arr;
  }

  public static int[] castToInts(Object arr) {
    return (int[]) arr;
  }

  public static long[] castToLongs(Object arr) {
    return (long[]) arr;
  }

  public static byte[] byteArrayForSize(Number size) {
    return new byte[size.intValue()];
  }

  public static short[] shortArrayForSize(Number size) {
    return new short[size.intValue()];
  }

  public static int[] intArrayForSize(Number size) {
    return new int[size.intValue()];
  }

  public static long[] longArrayForSize(Number size) {
    return new long[size.intValue()];
  }

  public static float[] floatArrayForSize(Number size) {
    return new float[size.intValue()];
  }

  public static double[] doubleArrayForSize(Number size) {
    return new double[size.intValue()];
  }

  public static boolean[] booleanArrayForSize(Number size) {
    return new boolean[size.intValue()];
  }

  public static char[] charArrayForSize(Number size) {
    return new char[size.intValue()];
  }

}

File: src/jvm/clojure/lang/primitive_array.clj
(ns clojure.lang.primitive-array
  (:refer-clojure :only [defmacro defn])
  (:require [clojure.next           :refer :all]
            [clojure.lang.protocols :refer [ISeq]])
  (:import  [clojure.lang.platform PrimitiveArray]))

(def platform-boolean        Boolean)
(def platform-native-boolean Boolean/TYPE)
(def platform-char           Character)

(defmacro to-booleans [arr]
  `(PrimitiveArray/castToBooleans ~arr))

(defmacro to-bytes [arr]
  `(PrimitiveArray/castToBytes ~arr))

(defmacro to-chars [arr]
  `(PrimitiveArray/castToChars ~arr))

(defmacro to-shorts [arr]
  `(PrimitiveArray/castToShorts ~arr))

(defmacro to-floats [arr]
  `(PrimitiveArray/castToFloats ~arr))

(defmacro to-doubles [arr]
  `(PrimitiveArray/castToDoubles ~arr))

(defmacro to-ints [arr]
  `(PrimitiveArray/castToInts ~arr))

(defmacro to-longs [arr]
  `(PrimitiveArray/castToLongs ~arr))

(defn byte-array-for-size [size]
  (PrimitiveArray/byteArrayForSize ^Number size))

(defn short-array-for-size [size]
  (PrimitiveArray/shortArrayForSize ^Number size))

(defn int-array-for-size [size]
  (PrimitiveArray/intArrayForSize ^Number size))

(defn long-array-for-size [size]
  (PrimitiveArray/longArrayForSize ^Number size))

(defn float-array-for-size [size]
  (PrimitiveArray/floatArrayForSize ^Number size))

(defn double-array-for-size [size]
  (PrimitiveArray/doubleArrayForSize ^Number size))

(defn boolean-array-for-size [size]
  (PrimitiveArray/booleanArrayForSize ^Number size))

(defn char-array-for-size [size]
  (PrimitiveArray/charArrayForSize ^Number size))

